### PR TITLE
8217633: Configurable extensions with system properties

### PR DESCRIPTION
**Pull request by XueleiFan**
The TLS protocols are designed to tolerant unknown TLS extensions. However, although it is not common, there are a few TLS implementations that cannot handle unknown extensions properly. As results in unexpected interoperability issue when new extensions are introduced in JDK. The interoperability impact could be mitigated If applications can customize the extensions if needed.

With this update, two system properties are added to configure the default extensions in either client or server side of TLS connections.  Please note that the impact of blocking TLS extensions is complicated.  For example, a TLS connection may not be able to established if a mandatory extension is blocked.  Please don't use this feature unless you clearly understand the impact.

Bug: https://bugs.openjdk.java.net/browse/JDK-8217633 
CSR: https://bugs.openjdk.java.net/browse/JDK-8217993

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8217633](https://bugs.openjdk.java.net/browse/JDK-8217633): Configurable extensions with system properties

### Reviewers
 * [Rajan Halade](https://openjdk.java.net/census#rhalade) (@rhalade - **Reviewer**) ⚠️ Review applies to 5bd6e865554af9a4cc0791fe4b9cb269699a0629
 * [Jamil Nimeh](https://openjdk.java.net/census#jnimeh) (@jnimeh - **Reviewer**) ⚠️ Review applies to ad5f3330b873ab36e6354a3cea14a87aefa0c083

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1752/head:pull/1752`
`$ git checkout pull/1752`

### CODE REVIEW

**Comment by XueleiFan** [comment]
ping ...

**Review by RogerRiggs — COMMENTED** [review]
Where will the new properties be documented?

In the CSR, it may be worth mentioning that the extension names are case sensitive.

**Comment by XueleiFan** [comment]
> Where will the new properties be documented?
> 
Yes.  The properties will be described in the JSSE Reference Guide.  I added a "Documentation" section in the CSR.

> In the CSR, it may be worth mentioning that the extension names are case sensitive.
Good point.

Thanks for the review.

File: test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2020, Oracle and/or its affiliates. All rights reserved.
```
Comment (rhalade) [inline test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java:2]:
```suggestion
 * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
```

File: test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    public static void main(String[] args) throws Exception {
+        System.setProperty("jdk.tls.client.blockedExtensions", args[0]);
+
+        boolean shouldSuccess = (args.length != 3);
+
+        try {
+            (new BlockedExtension(new String[] {args[1]})).run();
+        } catch (SSLException | IllegalStateException ssle) {
+            if (shouldSuccess) {
+                throw new Exception(
```
Comment (rhalade) [inline test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java:62]:
```suggestion
                throw new RuntimeException(
```

File: test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                        "The extension " + args[0] + " is blocked");
+            }
+
+            return;
+        }
+
+        if (!shouldSuccess) {
+            throw new Exception(
```
Comment (rhalade) [inline test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java:70]:
```suggestion
            throw new RuntimeException(
```

**Review by rhalade — CHANGES_REQUESTED** [review]

File: test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java (hunk 3 above)
Comment (XueleiFan) [reply test/jdk/sun/security/ssl/SSLSocketImpl/BlockedExtension.java]: Good catches, the years should be 2021 now.  Thanks!

**Review by jnimeh — APPROVED** [review]
Looks good to me.

**Review by rhalade — APPROVED** [review]

**Comment by XueleiFan** [comment]
Hi Bernd,

I agree with you that System property is not as useful to configure individual connections.  It is mostly used for corner cases that have interoperability or compatibility issues.  A general program should use APIs and the default system properties. 

> _Mailing list message from [Bernd Eckenfels](mailto:ecki@zusammenkunft.net) on [security-dev](mailto:security-dev@openjdk.java.net):_
> 
> Hello,
> 
> I wanted to mention again, that all those System property configurations are good, especially to resolve the update pains, but not really useful if you want to make configurations on a per-connection base. If you have to support multiple partners it can be a real pain to setup a common feature set or multiple instances. For this a generic feature setter for the context would be really useful. Most prominent recent example is the ca-extension, which only really makes sense if you also did programmatically configure a small list of trusted CAs.
> 
Yes,  ca-extension is an item I was thinking of to support in JDK.

> I also think it would overall clean up the code and give a good place for Javadoc all those options.
> Not to mention the default could be tied to a few new context names.
> 
Currently, the system properties are documented in the JSSE Reference Guides.  But just as you know, it is as easy to follow.  I agree with you that it would be nice to have better place to have them all together.

Thank you for the review.

Regards,
Xuelei


> Gruss
> Bernd
> --
> http://bernd.eckenfels.net

**Comment by mdameda** [comment]
Hi @XueleiFan , i don't find this patch into latest OpenJDK 1.8 releases . Is there a plan to integrate this patch also into OpenJDK 1.8  ? (i see that OracleJDK 1.8 was updated)

**Pull request closed without merging**